### PR TITLE
Add the ability to enable and disable plugins while running - part of #682

### PR DESCRIPTION
**Pull request by peternewman**
Initial work to enable/disable plugins while running, the C++ client and RPC bits. Still need to actually do the work on the server side and implement in the HTTP/JSON and Python APIs

### CODE REVIEW

File: common/protocol/Ola.proto [hunk 1]
```diff
@@ -403,6 +406,7 @@ service OlaServerService {
   rpc GetDeviceInfo (DeviceInfoRequest) returns (DeviceInfoReply);
   rpc GetCandidatePorts (OptionalUniverseRequest) returns (DeviceInfoReply);
   rpc ConfigureDevice (DeviceConfigRequest) returns (DeviceConfigReply);
+  rpc SetPluginState (PluginStateChange) returns (Ack);
```
Comment (nomis52) [inline common/protocol/Ola.proto]: I'd call this PluginStateRequest

Comment (peternewman) [inline common/protocol/Ola.proto]: I can't, it already exists to get the plugin info; see line 402/405.

Comment (nomis52) [inline common/protocol/Ola.proto]: How about 

PluginStateChangeRequest

I know it's long, but it's much easier to keep track of the types if they have a Request / Reply suffix.

Comment (peternewman) [inline common/protocol/Ola.proto]: Done.

**Comment by nomis52** [comment]
It looks like this is pretty close to being done. Do you need anything else from me?

**Comment by peternewman** [comment]
Not yet, I need to add the code to actually do the work in olad and expose it via the HTTP interface.

**Comment by peternewman** [comment]
@nomis52 this should be ready for review and merge now.

File: examples/ola-client.cpp [hunk 2]
```diff
@@ -278,6 +279,12 @@ void DisplayDevices(SelectServer *ss,
   ss->Terminate();
 }
 
+void SetPluginStateComplete(SelectServer *ss, const Result &result) {
```
Comment (nomis52) [inline examples/ola-client.cpp]: We should just merge all of these into a HandleAck(..) but that can be done later.

File: examples/ola-client.cpp [hunk 3]
```diff
@@ -813,8 +826,23 @@ int FetchPluginState(OlaClientWrapper *wrapper, const options &opts) {
     DisplayPluginStateHelp(opts);
     exit(1);
   }
-  client->FetchPluginState((ola::ola_plugin_id) opts.plugin_id,
-                           NewSingleCallback(&DisplayPluginState, ss));
+  if (opts.state.length() > 0) {
```
Comment (nomis52) [inline examples/ola-client.cpp]: !opts.state.empty()

File: include/ola/client/OlaClient.h [hunk 4]
```diff
@@ -130,6 +131,16 @@ class OlaClient {
                            CandidatePortsCallback *callback);
 
   /**
+   * @brief Set the state of a plugin.
+   * @param plugin_ids the ids of the plugins to set the state of.
+   * @param state the state to set the plugin to
+   * @param callback the SetCallback to invoke upon completion.
+   */
+  void SetPluginState(const std::vector<ola_plugin_id> &plugin_ids,
```
Comment (nomis52) [inline include/ola/client/OlaClient.h]: Make this take a single ID.

The problem with passing multiple plugins in a single call is how we handle failures. Do we change all or none? That means we have to handle rollbacks.  It gets complicated.

File: common/protocol/Ola.proto [hunk 5]
```diff
@@ -196,6 +196,11 @@ message PluginStateReply {
   repeated PluginInfo conflicts_with = 5;
 }
 
+message PluginStateChangeRequest {
+  repeated int32 plugin_id = 1;
```
Comment (nomis52) [inline common/protocol/Ola.proto]: Make this singular - see comments below.

File: olad/OladHTTPServer.cpp [hunk 6]
```diff
@@ -508,6 +513,44 @@ int OladHTTPServer::ModifyUniverse(const HTTPRequest *request,
...
+
+  string state_string = request->GetPostParameter("state");
+
+  bool state;
+  if (!StringToBoolTolerant(state_string, &state)) {
+    OLA_INFO << "Invalid state " << state_string;
+    return ServeHelpRedirect(response);
+  }
+
+  string plugin_id_string = request->GetPostParameter("plugin_ids");
```
Comment (nomis52) [inline olad/OladHTTPServer.cpp]: We can fan out here if we really need to support changing multiple plugins at once, but I'd avoid it if we can.

File: examples/ola-client.cpp (hunk 3 above)
Comment (peternewman) [inline examples/ola-client.cpp]: Done.

File: common/protocol/Ola.proto (hunk 5 above)
Comment (peternewman) [inline common/protocol/Ola.proto]: Done.

File: olad/OladHTTPServer.cpp (hunk 6 above)
Comment (peternewman) [inline olad/OladHTTPServer.cpp]: Removed.

File: include/ola/client/OlaClient.h (hunk 4 above)
Comment (peternewman) [inline include/ola/client/OlaClient.h]: Done.

File: examples/ola-client.cpp (hunk 2 above)
Comment (peternewman) [inline examples/ola-client.cpp]: Done.

File: ola/OlaClientCore.h [hunk 7]
```diff
@@ -23,6 +23,7 @@
 
 #include <memory>
 #include <string>
+#include <vector>
```
Comment (nomis52) [inline ola/OlaClientCore.h]: Can drop this now

File: include/ola/client/OlaClient.h [hunk 8]
```diff
@@ -34,6 +34,7 @@
 
 #include <memory>
 #include <string>
+#include <vector>
```
Comment (nomis52) [inline include/ola/client/OlaClient.h]: And here

File: ola/OlaClientCore.h (hunk 7 above)
Comment (peternewman) [inline ola/OlaClientCore.h]: Done

File: include/ola/client/OlaClient.h (hunk 8 above)
Comment (peternewman) [inline include/ola/client/OlaClient.h]: Done

**Pull request merged**